features/map.go

```diff
@@ -0,0 +1,173 @@
+	mc.mu.Lock()
+	delete(mc.mapTypes, mt)
+	mc.mu.Unlock()
+}
+
+func probeMapTypeAttr(mt ebpf.MapType) *internal.BPFMapCreateAttr {
+	var (
+		keySize               uint32 = 4
+		valueSize             uint32 = 4
+		maxEntries            uint32 = 1
+		innerMapFd            uint32
+		flags                 uint32
+		btfKeyTypeID          uint32
+		btfValueTypeID        uint32
+		btfFd                 uint32
+		btfVmLinuxValueTypeID uint32
+	)
+
+	// switch on map types to generate correct bpfMapCreateAttr
+	switch mt {
+	case ebpf.StackTrace:
+		// valueSize needs to be sizeof(uint64)
+		valueSize = 8
+	case ebpf.LPMTrie:
+		keySize = 8
+		valueSize = 8
+		flags = unix.BPF_F_NO_PREALLOC
+	case ebpf.ArrayOfMaps:
+		fallthrough
+	case ebpf.HashOfMaps:
+		innerMapFd = ^uint32(0)
+	case ebpf.CGroupStorage:
+		fallthrough
+	case ebpf.PerCPUCGroupStorage:
+		// Why struct{u32 + u64} = 12 + padding = 16 ? can it be 12 for 32-bit cpus?
+		keySize = 16
+		valueSize = 8
+		maxEntries = 0
+	case ebpf.Queue:
+		fallthrough
+	case ebpf.Stack:
+		keySize = 0
+	case ebpf.StructOpts:
+		// we can't support StructOps probes currently as it will require a valid BTF fd
+		btfVmLinuxValueTypeID = 1
+	case ebpf.RingBuf:
+		keySize = 0
+		valueSize = 0
+		maxEntries = 4096
+	case ebpf.SkStorage:
+		fallthrough
+	case ebpf.InodeStorage:
+		fallthrough
+	case ebpf.TaskStorage:
+		valueSize = 8
+		maxEntries = 0
+		flags = unix.BPF_F_NO_PREALLOC
+		btfKeyTypeID = 1
+		btfValueTypeID = 3
+		btfFd = ^uint32(0)
+	}
+
+	return &internal.BPFMapCreateAttr{
+		MapType:               uint32(mt),
+		KeySize:               keySize,
+		ValueSize:             valueSize,
+		MaxEntries:            maxEntries,
+		InnerMapFd:            innerMapFd,
+		Flags:                 flags,
+		BTFKeyTypeID:          btfKeyTypeID,
+		BTFValueTypeID:        btfValueTypeID,
+		BTFFd:                 btfFd,
+		BTFVmLinuxValueTypeID: btfVmLinuxValueTypeID,
+	}
+
+}
+
+// ProbeMapType allows probing the availability of a specified ebpf.MapType
+// It will call the syscall to create a dummy map of a given ebpf.MapType at most once
+// storing the result of the first call in a global cache.
+// This potentially can result in false results if the calling process changes its permissions
+// or capabilities.
+// Calling programs can avoid false cached results by invalidating the cache through
+// FlushMapCache() and FlushMapCacheEntry().
+func ProbeMapType(mt ebpf.MapType) error {
+	// make sure to bound Map types
+	// MaxMapType new value in enum, easier to handle than making sure
+	// we are checking the last value in the enum (which could eventually change)
+	if mt >= ebpf.MaxMapType {
+		return internal.ErrNotSupported
+	}
+
+	mc.mu.RLock()
+	if err, ok := mc.mapTypes[mt]; ok {
+		defer mc.mu.RUnlock()
+		return err
+	}
+	mc.mu.RUnlock()
+
+	attr := probeMapTypeAttr(mt)
+	_, err := internal.BPFMapCreate(attr)
+
+	// For nested and storage maps we accept EBADF as indicator that nested maps are supported
```

Comment:
Yup, agree with the reasoning overall. It seems like we cannot create StructOps maps without vmlinux BTF available, hence we could already return ErrNotSupported if a valid BTF blob cannot be found.

> If it is present, then we can maybe open a FD on it

Maybe I didn't get this right, but is loading the kernel's own vmlinux BTF blob into the kernel necessary, though? Doesn't the kernel already have the ability to introspect its own types? All that's needed to determine `btfVmLinuxValueTypeID` is parsing the BTF blob in userspace and querying it for the `struct bpf_struct_ops_tcp_congestion_ops` type. After all, isn't that what `btf__find_by_name_kind(btf_vmlinux, "bpf_struct_ops_tcp_congestion_ops", ...` does in your example in #344? `libbpf_find_kernel_btf()` doesn't return an fd.

Reply:
Not sure I understand.

> Maybe I didn't get this right, but is loading the kernel's own vmlinux BTF blob into the kernel necessary (or even possible), though? Doesn't the kernel already have the ability to introspect its own types?

As far I understand the struct_ops maps rely on the availability of the vmlinux BTF in the kernel. So yes, the vmlinux BTF is loaded into the kernel somehow? But I don't think _we_ should attempt to load it again from user space. It should be possible to retrieve a file descriptor to that BTF object if it's present in the kernel, with `bpf_btf_get_fd_by_id()`? It still requires a way to find the id for kernel BTF, though :thinking:.

Then once we have this fd we could assign it to `attr->btf_fd`, this won't be valid and the map won't be created, but it should return `-EBADF` instead of `-EINVAL`, so we could conclude that struct_ops maps are supported.

Reply:
You're definitely on to something, let's resume this in https://github.com/cilium/ebpf/issues/344!